gem/impero_utils.py

```diff
@@ -88,7 +96,26 @@ def nonzero(assignment):
     declare, indices = place_declarations(tree, temporaries, get_indices)
 
     # Prepare ImperoC (Impero AST + other data for code generation)
-    return ImperoC(tree, temporaries, declare, indices)
+    return ImperoC(tree, temporaries, declare, indices, return_variable)
+
+
+def infer_dtype(assignments, scalar_type):
+    from tsfc.loopy import assign_dtypes
```

Comment:
I know this is just a draft pull request, but this line would create a circularity in package dependency, by making `gem` depend on `tsfc`.  This in undesirable: `gem` should not depend on anything other than the Python standard library and quasi-standard packages.

Reply:
Yeah, that's easy to fix by doing the more general gem typing approach.